framework/include/materials/Material.h

```diff
@@ -48,7 +48,7 @@ class Material : public MooseObject,
                  public BlockRestrictable,
                  public BoundaryRestrictable,
                  public SetupInterface,
-                 public Coupleable,
+                 public NeighborCoupleable,
```

Comment:
If you do this, you are saying that in a Material values from a neighbor are available. In a volumetric instance of the material, (a) there are no neighbors, (b) reinit neighbor was not called. So you just gave the user an uninitialized memory.  The same holds for the face version of the material.

The only time you could do this is in the neighbor variant of the material. So my question is, why do you need neighbor variable values in a material? Materials were not really build for this. Why would not you use `InternalSideUserObject`s instead?

Reply:
Dear David,
I see your point about conceptual material neighbour.
I didn't use an `InternalSideUserObjects` because I haven't thought about it.
If I understand correctly you are suggesting to use the `InternalSideUserObjects` to grab material properties and variables on both side of the interface. This UO will then be called from the `boundary restricted material`, that will call the `traction separation UO`. 
The `boundary restricted material` will be used only to store the interface stateful property and residual and Jacobian coefficients required by the `interface kernel`.

Is this correct? Can we avoid using two UO or it makes more sense to keep them separate?

Also right now the `traction separation UO` is a `SideUserObject`

Reply:
@andrsd We use a stateful material to store the displacement jump across the interface and to get the jump we need to know the neighbor variable values. If use ``InternalSideUserObject``, does that mean we have to manually store the stateful jump values at each qp and each element?

Reply:
@arovinelli Yes, `InternalSideUO` should give you variable values and material values for both sides. Then, you compute whatever you need to and store it.  Then, you would have a special API to get that info back from where ever you needed it.

@jiangwen84 Yes, you would have to take care of the storage yourself.

Reply:
@andrsd why should I need a special API? Can't I simply do the same I do for the traction separation law, compute interface related quantities in the `InternalSideUO` and directly save them into a material property?

Reply:
User objects do not write into material properties. That's what materials do, because only those have writable references for that memory. Every other system gets a const reference. Cannot write into those.   So, if you have a need to store stuff into material properties, you write a material, that will extract that info from a user object via a special API.

Reply:
@jiangwen84 and @andrsd 
can you please take a look at this 
 [czm_rework](https://github.com/arovinelli/moose/compare/czm_DG_%2311456...arovinelli:czm_reworking)
and tell me if this is what you had in mind?

I changed back the material to `coupleable`, compute the displacement jump on `InternalSideUO`, stored the jump in the interface and used the rest of the code

There is test already a working

Reply:
I do not think this works. When you call your UO from a Material, the neighbor values were not computed. Recall what I said earlier "on the volumetric and face version of a material, neighbor values are not computed". So your UO is using some garbage values . Most likely values from the last evaluated neighbor.  That's why you have to store the values first in the user object and then get them from wherever you need them.

Reply:
@andrsd this make sense, could you point me to an example or a test doing the storage and the API access? I guess with that in hand it would be an easy fix

Reply:
@andrsd I think the tricky part is how to store the stateful material property. Do you have any suggestions?

Reply:
Wait a second, I think I need to clarify some concept before going ahead. 
The `bulk element MP` have nothing to do with the `interface MP`.
An interface MP, like for instance a damage parameter `d`, are defined only on the interface and on one side because they are jumps.
However to compute an `interface MP` we need bulk `element MP`, like for instance the stress, and/or 'field variable' on both sides of the interface but coming from the bulk element. An example of the latter is the displacement which is used to compute the jump. Therefore the `interface MP` need to be compute either concurrently or after the bulk MP and field variables have been updated. (we may even lag `interface MP` of 1 NL iteration ).

With this being said we need simply to delay the computation of the interface material after all the 'bulk MP' and `field variables` have been updated. As a consequence this will also make the call to the UO at the appropriate time removing the need to store `interface MP` into the user object.
I guess this could also be achieved by postponing the call of the IK which requires `interface MP`

I hope this clarify the problem

Reply:
@arovinelli First, you do want the interface material be computing at every linear iteration because they are used in the residual calculation, that is updated for every linear iteration. Second, when you call UO from the material, the uo is not executed right away to do the calculation for the same element/side. UO is executed inside a separate loop. @andrsd please correct me if I am wrong.

Reply:
> Second, when you call UO from the material, the uo is not executed right away to do the calculation for the same element/side. UO is executed inside a separate loop. @andrsd please correct me if I am wrong.

This is correct.

I was thinking about your need of stateful material properties. Do you actually have a state that depends on the previous values (like in case where the state would be "accumulated" over time) or do you just need old values and you used stateful matprops to avoid recomputing them?

Reply:
@andrsd yes we do have a state that accumulates, which is the damage parameter on the interface. This is not implemented on the current version, but is the next step after we figure out a solution to the current problem.

Reply:
@andrsd I was thinking about what you said about computing the UO independently and storing all the values for the material.

Is what is done in [example 20](http://mooseframework.org/wiki/MooseExamples/Example_20/) on the line of what you were thinking at ?

Reply:
yes

Reply:
@andrsd I update [czm_rework](https://github.com/arovinelli/moose/compare/czm_DG_%2311456...arovinelli:czm_reworking) and did as suggested such that the `InteranlSideUO` takes care of grabbing MP and variables on both side of the interface,
so you may want to look at it.
The calculations for computing residual and jacobian coefficients are still done by the material calling directly the 'SideUO' . Is this ok, or I need to pre-compute everything in the `Internal Side UO`. I am sorry about all these question but the order of execution is still confusing for me (by the way is there any good reference for it?).

Reply:
Dear @andrsd adn @jiangwen84 .
I changed a bit the design and moved all the computation for the czm in the `InternalSideUO`.
Now the boundary material just probes the data stored in InternalSideUO and computes required quantity on the fly (at least I believe). Let me explain, for each qp the UO only store basic quantities (like the `displacement jump` and the `normal`) and when the get method is called, Residual and Jacobian coefficients are computed. Now before moving even further I would like to know if you think this is what we really want and that required values are in sync with what we are doing.
All the best